### PR TITLE
Fix trace sdk builder 1393

### LINKED ISSUE #1393: [Trace SDK] Provide builders to avoid exposing SDK internals
Greetings.

In an application, as a **_user_** of the trace sdk, I need to create a `opentelemetry::exporter::otlp::OtlpGrpcExporter` instance.

Per instructions in `exporters/otlp/README.md`,
the only way to do this seems to call new on the class itself, which imply that:
- the code `#include <opentelemetry/exporters/otlp/otlp_grpc_exporter.h>`
- the user code compiled in the application now depends on the exact memory layout of `OtlpGrpcExporter`.

This is problematic for two reasons:

1. In the specific case of `OtlpGrpcExporter`, the header file `otlp_grpc_exporter.h` contains:
`#include "opentelemetry/proto/collector/trace/v1/trace_service.grpc.pb.h"`

To build my application, I now need to compile this header from protobuf, and I need to install all the dependencies that are generated as well. This is very inconvenient, but still possible. It will slow down adoption of opentelemetry-cpp in any case.

2. In general, the content of `otlp_grpc_exporter.h` is very dependent on the SDK implementation, and is likely to change, causing compatibility issues with the user code. Delivering fixes without breaking the application is now much more difficult, which is the rationale for filing this as a bug and not a feature request.

Now, taking a step back, for applications merely **_using_** the SDK, they do not need to actually know the details of an exporter implementation, they only needs an `opentelemetry::sdk::trace::SpanExporter` to build a TraceProvider.

Applications **_extending_** the SDK will still need to see more details, this is a different topic.

I would suggest to create Builder classes, like:

```
// In namespace opentelemetry::exporter::otlp::
class OtlpGrpcExporterBuilder {
public:
  static std::unique_ptr<opentelemetry::sdk::trace::SpanExporter>
  Build(opentelemetry::exporter::otlp::OtlpGrpcExporterOptions options);
};

```
And implement the builder in a .cc file (in the exporter library), not inline in the header, even for a one-liner.

Note how the builder returns `opentelemetry::sdk::trace::SpanExporter`, not `opentelemetry::exporter::otlp::OtlpHttpExporter`.

This resolves:
- the header file dependencies, as file `otlp_grpc_exporter.h` is no longer needed (options are in a separate header)
- the binary layout dependency on the SDK implementation classes, as class `OtlpHttpExporter` is never seen.

This pattern can be applied to every subclass of SpanExporter, Processor, etc, which are at the top level of the SDK.

This can be applied to Traces, Metrics, and Logs.

Regards.

### LINKED ISSUE #1473: [Examples] Memory ownership issues in InMemorySpanExporter usage example.
In example `multi_processor/main.cc`, the code:

1)

Builds a `InMemorySpanExporter` object

```
  InMemorySpanExporter *memory_span_exporter = new InMemorySpanExporter();
  auto exporter2 = std::unique_ptr<trace_sdk::SpanExporter>(memory_span_exporter);

```
2)

Transfers memory ownership to a `SpanProcessor`

```
  auto processor2 = std::unique_ptr<trace_sdk::SpanProcessor>(
      new trace_sdk::SimpleSpanProcessor(std::move(exporter2)));

```

3)

Use the underlying `InMemorySpanExporter` object directly.

```
  InMemorySpanExporter *memory_span_exporter = initTracer();

  foo_library();
  auto memory_spans = memory_span_exporter->GetData()->GetSpans();
```

This usage of `InMemorySpanExporter` is unsafe, as the exporter pointer is not owned, and the exporter could go away for example when installing a different global tracer provider (if I understand correctly).

Suggested fix:

`initTracer()` should return the underlying `InMemorySpanData` instead, which is safe to do because the data is held by a shared pointer.
Even if the exporter goes away, the data will still be accessible.

### PR DESCRIPTION
**Pull request by marcalff**
(Last edit 2022-07-12)

Fixes #1393
Fixes #1473

## Changes

Implement Trace SDK builders, to avoid exposing SDK details to the application.
This is needed to deploy opentelemetry-cpp as a shared library.

Fixed memory ownership of `InMemorySpanExporter` in examples.

For significant contributions please make sure you have completed the following items:

* [X] `CHANGELOG.md` updated for non-trivial changes
* [X] Unit tests have been added
* [X] Changes in public API reviewed (no API change)
* [X] Added design documentation

### CODE REVIEW

**Comment by marcalff** [comment]
This PR is now ready for review.

Remaining tasks:
- run again all examples locally
- add doxygen comments
- write Changelog

Reviewers:
- Please start by reading docs/cpp-sdk-factory-design.md
- Please check the Bazel BUILD files in particular (I never used Bazel)

Regards,
-- Marc

**Review by lalitb — APPROVED** [review]
Very nicely done. Thanks for the PR. Looks good with few (very) minor comments. Hope it would be fine to merge @owent 's async changes first :)

File: docs/cpp-sdk-factory-design.md
Comment (marcalff) [reply docs/cpp-sdk-factory-design.md]: Thanks @lalitb, and agree.

Revised the wording in general.

File: examples/batch/main.cc
Comment (marcalff) [reply examples/batch/main.cc]: Fixed, cleaned up includes that got affected by this patch.

I agree, we need to clean up in includes in general by running include-what-you-use,
for the entire repository (in a separate PR).

File: exporters/memory/CMakeLists.txt
Comment (marcalff) [reply exporters/memory/CMakeLists.txt:1]: Before the change, "opentelemetry_exporter_in_memory" was in fact a header only SDK,
no actual library was built and installed, so I don't think there are risks of breaking anything.

With the change, the memory exporter factory needs to be in a *.cc file,
and so we need a (real) library for it.

I used the name opentelemetry_exporter_memory_trace,
to allow for metrics memory exporter libraries later.

**Comment by marcalff** [comment]
> Very nicely done. Thanks for the PR. Looks good with few (very) minor comments. Hope it would be fine to merge @owent 's async changes first :)

Thanks.

The async changes have priority, I will deal with any merge conflict, these should be trivial to resolve anyway.

One remaining point to discuss, that I am not sure of, is whether Factory classes should use:
- std::shared_ptr
- nostd::shared_ptr
because this is part of the application owner / SDK interface.

@lalitb , @owent , @ThomsonTan , @esigo any input ?

**Comment by owent** [comment]
> > Very nicely done. Thanks for the PR. Looks good with few (very) minor comments. Hope it would be fine to merge @owent 's async changes first :)
> 
> Thanks.
> 
> The async changes have priority, I will deal with any merge conflict, these should be trivial to resolve anyway.
> 
> One remaining point to discuss, that I am not sure of, is whether Factory classes should use:
> 
> * std::shared_ptr
> * nostd::shared_ptr
>   because this is part of the application owner / SDK interface.
> 
> @lalitb , @owent , @ThomsonTan , @esigo any input ?

Or `std::unique_ptr` so users can convert it to any implementation of `shared_ptr`?

**Comment by marcalff** [comment]
> Or `std::unique_ptr` so users can convert it to any implementation of `shared_ptr`?

Thanks, good point: a factory method should return a unique_ptr anyway.
Fixed.

**Comment by lalitb** [comment]
@owent Can you please review/approve it once you have time? Thanks :)

**Comment by owent** [comment]
> @owent Can you please review/approve it once you have time? Thanks :)

Sorry for something else in company take my bandwidth. I will review this it in 2 or 3 days.

**Comment by lalitb** [comment]
> > @owent Can you please review/approve it once you have time? Thanks :)
> 
> Sorry for something else in company take my bandwidth. I will review this it in 2 or 3 days.

Thanks. That should be fine :)

**Comment by owent** [comment]
@marcalff Cloud you add a standalone unit test which only include factory headers and do not depend headers of protobuf? I think it's helpful to keep factory heads clean in the future.

File: exporters/otlp/src/otlp_grpc_exporter_factory.cc
Comment (owent) [inline exporters/otlp/src/otlp_grpc_exporter_factory.cc:20]: Should we use `Create(OtlpGrpcExporterOptions{})` to reduce copy here?

File: exporters/otlp/src/otlp_http_exporter_factory.cc
Comment (owent) [inline exporters/otlp/src/otlp_http_exporter_factory.cc:17]: Should we use `Create(OtlpHttpExporterOptions{})` to reduce copy here?

File: exporters/zipkin/src/zipkin_exporter_factory.cc
Comment (owent) [inline exporters/zipkin/src/zipkin_exporter_factory.cc:19]: Just like OTLP exporters, should we use `Create(ZipkinExporterOptions{})` to reduce copy here?

File: sdk/src/trace/tracer_provider_factory.cc
Comment (owent) [inline sdk/src/trace/tracer_provider_factory.cc:57]: Should we use `Create(std::move(processors), opentelemetry::sdk::resource::Resource::Create({}));` or `Create(std::move(processors), std::move(resource));` to reduce copy?

File: sdk/src/trace/tracer_provider_factory.cc
Comment (owent) [inline sdk/src/trace/tracer_provider_factory.cc:22]: Should we use `Create(std::move(processors), opentelemetry::sdk::resource::Resource::Create({}));` or `Create(std::move(processors), std::move(resource));` to reduce copy?

File: sdk/src/trace/tracer_provider_factory.cc
Comment (owent) [inline sdk/src/trace/tracer_provider_factory.cc:30]: Should we use `Create(std::move(processors), std::move(resource), std::move(sampler));` to reduce copy?

**Review by owent — COMMENTED** [review]
Could you please add a standalone unit test for OTLP exporters and ES exporters which only include factory headers and do not depend headers of protobuf? I think it's helpful to keep factory heads clean in the future.

File: exporters/otlp/src/otlp_grpc_exporter_factory.cc
Comment (marcalff) [reply exporters/otlp/src/otlp_grpc_exporter_factory.cc:20]: Thanks for the comments.
Here the Create() method with an option takes a `const OtlpGrpcExporterOptions &options` parameter,
so there should be no copy.

File: exporters/otlp/src/otlp_http_exporter_factory.cc
Comment (marcalff) [reply exporters/otlp/src/otlp_http_exporter_factory.cc:17]: Likewise, the Create() method with an option takes a `const &` parameter,
so there should be no copy.

File: exporters/zipkin/src/zipkin_exporter_factory.cc
Comment (marcalff) [reply exporters/zipkin/src/zipkin_exporter_factory.cc:19]: Likewise, the Create() method with an option takes a `const &` parameter,
so there should be no copy.

File: sdk/src/trace/tracer_provider_factory.cc
Comment (marcalff) [reply sdk/src/trace/tracer_provider_factory.cc:57]: Here there is a copy, thanks for pointing this out.

Fixed by using `const opentelemetry::sdk::resource::Resource &resource` in every` Create()` method.

File: sdk/src/trace/tracer_provider_factory.cc
Comment (marcalff) [reply sdk/src/trace/tracer_provider_factory.cc:22]: Likewise,

Fixed by using `const opentelemetry::sdk::resource::Resource &resource` in every` Create()` method.

File: sdk/src/trace/tracer_provider_factory.cc
Comment (marcalff) [reply sdk/src/trace/tracer_provider_factory.cc:30]: Likewise,

Fixed by using `const opentelemetry::sdk::resource::Resource &resource` in every` Create()` method.

File: exporters/otlp/test/otlp_grpc_exporter_factory_test.cc
Comment (owent) [inline exporters/otlp/test/otlp_grpc_exporter_factory_test.cc:16]: Maybe we can add a `static_assert` to ensure we will not break the header dependency in the future?

```cpp
#ifdef GOOGLE_PROTOBUF_VERSION
static_assert(false, "protobuf should not be included here");
#endif
```

Comment (owent) [inline exporters/otlp/test/otlp_grpc_exporter_factory_test.cc:16]: Maybe we can also add a `static_assert` to ensure we will not break the header dependency in the future?

```cpp
#ifdef GOOGLE_PROTOBUF_VERSION
#  error  "protobuf should not be included"
#endif
```

File: exporters/otlp/test/otlp_http_exporter_factory_test.cc
Comment (marcalff) [reply exporters/otlp/test/otlp_http_exporter_factory_test.cc:16]: Fixed as indicated.

The build error message is actually better, thanks.

Not sure what you meant by 'static_assert', so please clarify if something else is needed.

File: exporters/otlp/test/otlp_grpc_exporter_factory_test.cc
Comment (marcalff) [reply exporters/otlp/test/otlp_grpc_exporter_factory_test.cc:8]: Thanks, fixed.

**Review by owent — APPROVED** [review]
LGTM

**Comment by marcalff** [comment]
@lalitb, @ThomsonTan  Do not merge yet.

It turns out the library name for the in memory exporter is visible externally after all,
in `cmake/opentelemetry-cpp-config.cmake.in`

`  opentelemetry-cpp::in_memory_span_exporter        - Imported target of opentelemetry-cpp::in_memory_span_exporter`

Reverting un needed changes in `exporter/memory/CMakeLists.txt`, will notify when ready to merge.

File: exporters/memory/CMakeLists.txt
Comment (marcalff) [reply exporters/memory/CMakeLists.txt]: Reconsidered, and reverted the library name change.

The name is visible externally in `cmake/opentelemetry-cpp-config.cmake.in`

Comment (owent) [reply exporters/memory/CMakeLists.txt]: `cmake/opentelemetry-cpp-config.cmake.in` only use the `EXPORT_NAME` attribute, do not use internal target name.
Internal target name change will only break the projects use opentelemery-cpp as a submodule.

File: docs/cpp-sdk-factory-design.md
Comment (ThomsonTan) [inline docs/cpp-sdk-factory-design.md:36]: nit
```suggestion
application must be rebuilt (with a fixed version), redistributed,
```

Comment (marcalff) [reply docs/cpp-sdk-factory-design.md]: Thanks, fixed.

**Pull request merged**
